### PR TITLE
Stream version for Text2Ngram

### PR DESCRIPTION
**Pull request by Davidsirui**
### Which issue(s) does this Pull Request fix?
resolves #000

### CODE REVIEW

**Comment by jhfjhfj1** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- autokeras/tuner.py  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/autokeras/autokeras/pullRequest?prid=3920400&bid=13622031)

**Comment by abgese** [comment]
The code currently assumes 1-gram
Should I change it to take n-gram range as input?

**Comment by jhfjhfj1** [comment]
@abgese yes please. Thanks

File: autokeras/hypermodel/preprocessor.py [hunk 1]
```diff
@@ -144,65 +144,156 @@ def set_weights(self, weights):
 class TextToNgramVector(base.Preprocessor):
     """Convert raw texts to n-gram vectors."""
 
-    def __init__(self, **kwargs):
+    def __init__(self, ngram_range=(1, 1), stop_words=None, **kwargs):
```
Comment (jhfjhfj1) [inline autokeras/hypermodel/preprocessor.py]: You can follow the TextToIntSequence to add docstrings to this class to describe the two newly added args.

File: autokeras/hypermodel/preprocessor.py [hunk 2]
```diff
@@ -144,65 +144,156 @@ def set_weights(self, weights):
...
+            space_join = " ".join
+
+            for n in range(min_n, min(max_n + 1, n_original_tokens + 1)):
+                for i in range(n_original_tokens - n + 1):
+                    tokens_append(space_join(original_tokens[i: i + n]))
+
+        return tokens
 
     def update(self, x, y=None):
         # TODO: Implement a sequential version fit for both
```
Comment (jhfjhfj1) [inline autokeras/hypermodel/preprocessor.py]: This TODO can be removed.

File: autokeras/hypermodel/preprocessor.py [hunk 3]
```diff
@@ -144,65 +144,156 @@ def set_weights(self, weights):
...
+            except ValueError:
+                continue
+            if num - 1 <= self._max_features:
+                self.mask[num] = 1
+
+        self.result = self.mask * self.kbestfeature_value
+        # Refresh the mask&temp_vec for next time usage.
+        self.mask = np.zeros(self._max_features, dtype=int)
+        self.temp_vec = set()
+        # TODO: For each x, what is the type of return value?
```
Comment (jhfjhfj1) [inline autokeras/hypermodel/preprocessor.py]: The output should be an eager tensor. The input x is also an eager tensor. The self.result is a local variable. Is it used anywhere else? if not we can make it local.

File: autokeras/hypermodel/preprocessor.py [hunk 4]
```diff
@@ -144,65 +144,156 @@ def set_weights(self, weights):
...
+        return self.result
 
     def output_types(self):
-        return (tf.float32,)
+        return tf.float64,
```
Comment (jhfjhfj1) [inline autokeras/hypermodel/preprocessor.py]: Please use "(tf.float64,)" which make it clearer that it is a tuple.

File: tests/autokeras/hypermodel/preprocessor_test.py [hunk 5]
```diff
@@ -72,12 +73,27 @@ def test_ngram():
...
+        common.generate_data(dtype='dataset'),
+        tf.float32)
+    assert isinstance(new_dataset, tf.data.Dataset)
+    # Test stop word removal
+    new_dataset = run_preprocessor(
+        preprocessor_module.TextToNgramVector(stop_words=ENGLISH_STOP_WORDS),
+        dataset,
+        common.generate_data(dtype='dataset'),
+        tf.float32)
+    assert isinstance(new_dataset, tf.data.Dataset)
```
Comment (jhfjhfj1) [inline tests/autokeras/hypermodel/preprocessor_test.py]: We should separate these assertions run_preprocessor calls to different test functions according to the single path rule of unittesting. Please add one more test to verify the output of your implementation is almost the same as the output of the original sklearn method.

File: autokeras/hypermodel/preprocessor.py (hunk 3 above)
Comment (abgese) [reply autokeras/hypermodel/preprocessor.py]: So should I use the tf.convert_to_tensor on the numpy array being returned?

File: autokeras/hypermodel/preprocessor.py [hunk 6]
```diff
@@ -142,67 +142,174 @@ def set_weights(self, weights):
...
+            raise ValueError(
+                    "max_features=%r, needs to be a positive integer"
+                    % self._max_features)
+        self._shape = None
+        self.vocabulary = {}  # Vocabulary(Increase with the inputs)
+        self.sentence_containers = {}  # Number of sentence that contains the token
+        self.word_count = {}  # Count of words across documents
+        self._idf_vec = {}  # IDF of all the tokens
+        self.stc_num = 0  # The number of all the sentences in the raw doc
+        self.kbestfeature_value = []
```
Comment (jhfjhfj1) [inline autokeras/hypermodel/preprocessor.py]: Use the snake case. Rename it to `k_best_feature_values`.

File: autokeras/hypermodel/preprocessor.py [hunk 7]
```diff
@@ -142,67 +142,174 @@ def set_weights(self, weights):
...
 
     @property
     def output_shape(self):
-        return self.shape
+        return self._shape
 
     def get_weights(self):
-        return {'vectorizer': self.vectorizer,
-                'selector': self.selector,
+        return {'selector': self.selector,
```
Comment (jhfjhfj1) [inline autokeras/hypermodel/preprocessor.py]: In `get_weights`, `set_weights`, `get_config`, `set_config`, all fields should be covered. Some self.** are not covered here. The ones that set by the user should be in `get_config` and `set_config`, e.g., norm, stop_words, ngram_range, max_features. The ones inferred from the data should be in `get_weights` and `set_weights`, e.g. shape, vocabulary.

File: autokeras/hypermodel/preprocessor.py (hunk 3 above)
Comment (jhfjhfj1) [reply autokeras/hypermodel/preprocessor.py]: No, it should be converted automatically. The current implementation is good.

File: autokeras/hypermodel/preprocessor.py [hunk 8]
```diff
@@ -142,67 +142,174 @@ def set_weights(self, weights):
...
-                k=min(self.vectorizer.max_features, data.shape[1]))
-            self.selector.fit(data, self.targets)
+        self.stc_num += 1  # idf smoothing
+        for word in self.sentence_containers:
+            idf = (np.log((self.stc_num)/(self.sentence_containers[word] + 1)) + 1)
+            self._idf_vec[word] = idf
+        if len(self.vocabulary) < self._max_features:
+            self._max_features = len(self.vocabulary)
+        # Sort based on term frequency to match sklearn TfidfVectorizer
+        kbestfeature = dict(sorted(self._idf_vec.items(),
```
[the remark this answers is not in the document]
Comment (abgese) [reply autokeras/hypermodel/preprocessor.py]: renamed to k_best_idfs to better represent behaviour

File: autokeras/hypermodel/preprocessor.py (hunk 6 above)
Comment (abgese) [reply autokeras/hypermodel/preprocessor.py]: renamed to k_best_idf_values to better represent behaviour

File: autokeras/hypermodel/preprocessor.py [hunk 9]
```diff
@@ -142,67 +142,186 @@ def set_weights(self, weights):
...
+                'ngram_range': self.ngram_range,
+                'max_features': self._max_features,
+                'stop_words': self.stop_words,
+                'norm': self.norm}
+
+    def set_config(self, config):
+        self.selector = config['selector']
+        self.targets = config['targets']
+        self._max_features = config['max_features']
+        if(self._max_features <= 0):
```
Comment (jhfjhfj1) [inline autokeras/hypermodel/preprocessor.py]: The set config is not called by the user. It is just for saving and loading the model. The value check is not necessary.

File: autokeras/hypermodel/preprocessor.py [hunk 10]
```diff
@@ -142,67 +142,186 @@ def set_weights(self, weights):
...
+            raise ValueError(
+                    "max_features=%r, needs to be a positive integer"
+                    % self._max_features)
+        self.ngram_range = config['ngram_range']
+        self.stop_words = config['stop_words']
+        self.norm = config['norm']
+        if(self.norm not in ('l1', 'l2', None)):
```
Comment (jhfjhfj1) [inline autokeras/hypermodel/preprocessor.py]: This one is also not needed.

**Comment by jhfjhfj1** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/jhfjhfj1/autokeras/pullRequest?prid=3920400&bid=13622031)

**Pull request merged**